File: src/main/java/com/epri/dss/common/impl/DSSClassImpl.java
package com.epri.dss.common.impl;

import com.epri.dss.common.impl.DSSGlobals;
import com.epri.dss.common.CktElement;
import com.epri.dss.common.DSSClass;
import com.epri.dss.general.DSSObject;
import com.epri.dss.general.impl.DSSObjectImpl;
import com.epri.dss.parser.impl.Parser;
import com.epri.dss.shared.CommandList;
import com.epri.dss.shared.HashList;
import com.epri.dss.shared.ObjectList;
import com.epri.dss.shared.impl.HashListImpl;
import com.epri.dss.shared.impl.ObjectListImpl;

/**
 * Base class for all DSS collection classes.
 * Keeps track of objects of each class, dispatches edits, etc
 *
 */
public class DSSClassImpl<T extends DSSObject> implements DSSClass<T> {

	private static com.epri.dss.common.DSSClasses DSSClasses;

	protected String className;

	/** Index of present active element */
	protected int activeElement;

	protected CommandList commandList;

	protected int activeProperty;

	protected HashList elementNameList;

	protected int numProperties;

	protected String[] propertyName, propertyHelp;

	protected int[] propertyIdxMap;

	/** Maps property to internal command number */
	protected int[] revPropertyIdxMap;

	protected int classType;

	protected ObjectList<T> elementList;

	/** When device gets renamed */
	protected boolean elementNamesOutOfSynch;

	protected boolean saved;

	public DSSClassImpl() {
		super();
		// init size and increment
		elementList = new ObjectListImpl<T>(20);
		propertyName = null;
		propertyHelp = null;
		propertyIdxMap = null;
		revPropertyIdxMap = null;

		activeElement = -1;
		activeProperty = -1;

		elementNameList = new HashListImpl(100);
		elementNamesOutOfSynch = false;
	}

	protected void finalize() throws Throwable {
		// get rid of space occupied by strings
		for (int i = 0; i < numProperties; i++) propertyName[i] = "";
		for (int i = 0; i < numProperties; i++) propertyHelp[i] = "";

		propertyName = new String[0];
		propertyHelp = new String[0];
		propertyIdxMap = new int[0];
		revPropertyIdxMap = new int[0];
		elementList = null;
		elementNameList = null;
		commandList = null;

		super.finalize();
	}

	public int newObject(String objName) {
		DSSGlobals.doErrorMsg(
				"Reached base class of DSSClass for device \"" +objName+ "\"",
				"N/A",
				"Should be overridden.", 780);
		return 0;
	}

	public void setActiveElement(int value) {
		if ((value > 0) && (value <= elementList.size())) {
			activeElement = value;
			DSSGlobals.activeDSSObject = elementList.get(activeElement);

			if (DSSGlobals.activeDSSObject instanceof DSSCktElement)
				DSSGlobals.activeCircuit.setActiveCktElement( (CktElement) DSSGlobals.activeDSSObject );
		}
	}

	public int getActiveElement() {
		return activeElement;
	}

	/**
	 * Uses global parser.
	 */
	public int edit() {
		DSSGlobals.doSimpleMsg("DSSClass.edit() called. Should be overriden.", 781);
		return 0;
	}

	public int init(int Handle) {
		DSSGlobals.doSimpleMsg("DSSClass.init() called. Should be overriden.", 782);
		return 0;
	}

	/**
	 * Used by newObject().
	 */
	protected int addObjectToList(T Obj) {
		// put it in this collection's element list.
		elementList.add(Obj);
		elementNameList.add( Obj.getName() );

		if (elementList.size() > 2 * elementNameList.getInitialAllocation())
			reallocateElementNameList();

		setActiveElement(elementList.size() - 1);
		return activeElement;  // return index of object in list
	}

	public boolean setActive(String ObjName) {
		boolean result = false;

		if (elementNamesOutOfSynch)
			resynchElementNameList();
		int idx = elementNameList.find(ObjName);
		if (idx > 0) {
			setActiveElement(idx);
			DSSGlobals.activeDSSObject = elementList.get(idx);
			result = true;
		}
		return result;
	}

	/**
	 * Find an obj of this class by name.
	 */
	public Object find(String ObjName) {
		Object result = null;
		if (elementNamesOutOfSynch)
			resynchElementNameList();

		int idx = elementNameList.find(ObjName);
		if (idx > 0) {
			setActiveElement(idx);
			result = elementList.get(idx);
		}
		return result;
	}

	/**
	 * Get address of active obj of this class.
	 */
	public Object getActiveObj() {
		if (activeElement > 0) {
			return elementList.get(activeElement);
		} else {
			return null;
		}
	}

	public String getFirstPropertyName() {
		activeProperty = 0;
		return getNextPropertyName();
	}

	public String getNextPropertyName() {
		String result;

		if (activeProperty <= numProperties) {
			result = propertyName[activeProperty];
		} else {
			result = "";
		}

		activeProperty += 1;

		return result;
	}

	/**
	 * Find property value by string.
	 */
	public int propertyIndex(String prop) {
		int result = 0;  // default result if not found
		for (int i = 0; i < numProperties; i++) {
			if (prop.equalsIgnoreCase(propertyName[i])) {
				result = propertyIdxMap[i];
				break;
			}
		}
		return result;
	}

	/**
	 * Add no. of intrinsic properties.
	 */
	protected void countProperties() {
		numProperties = numProperties + 1;
	}

	/**
	 * Add properties of this class to propName.
	 */
	protected void defineProperties() {
		activeProperty += 1;

		propertyName[activeProperty] = "like";
		propertyHelp[activeProperty] = "Make like another object, e.g.:" +
				DSSGlobals.CRLF + DSSGlobals.CRLF +
				"new capacitor.C2 like=c1  ...";
	}

	protected int classEdit(final Object activeObj, int paramPointer) {
		// continue parsing with contents of parser
		if (paramPointer > 0) {
//			DSSObject obj = (DSSObject) ActiveObj;
			switch (paramPointer) {
			case 1:
				makeLike(Parser.getInstance().makeString());  // like command
				break;
			}
		}
		return 0;
	}

	protected int makeLike(String objName) {
		DSSGlobals.doSimpleMsg("DSSClass.makeLike() called. Should be overriden.", 784);
		return 0;
	}

	public int getFirst() {
		int result = 0;
		if (elementList.size() == 0) {
			result = 0;
		} else {

			setActiveElement(1);
			DSSGlobals.activeDSSObject = elementList.get(0);
			if (DSSGlobals.activeDSSObject instanceof DSSCktElement) {
				DSSGlobals.activeCircuit.setActiveCktElement( (CktElement) DSSGlobals.activeDSSObject );
				result = activeElement;
			}
		}
		return result;
	}

	public int getNext() {
		int result = -1;
		if (activeElement > elementList.size()) {
			result = 0;
		} else {

			DSSGlobals.activeDSSObject = elementList.getNext();

			if (DSSGlobals.activeDSSObject instanceof DSSCktElement) {
				DSSGlobals.activeCircuit.setActiveCktElement( (CktElement) DSSGlobals.activeDSSObject );
				result = activeElement;
			}
		}
		activeElement += 1;

		return result;
	}

	/**
	 * Helper routine for building property strings.
	 *
	 * Using the addProperty function, you can list the properties here in the order you want
	 * them to appear when properties are accessed sequentially without tags. Syntax:
	 *
	 * addProperty(<name of property>, <index in the edit case statement>, <help text>);
	 */
	public void addProperty(String propName, int cmdMapIndex, String helpString) {
		activeProperty += 1;

		propertyName[activeProperty] = propName;
		propertyHelp[activeProperty] = helpString;
		// map to internal object property index
		propertyIdxMap[activeProperty] = cmdMapIndex;
		revPropertyIdxMap[cmdMapIndex] = activeProperty;
	}

	protected void allocatePropertyArrays() {
		propertyName = new String[numProperties];
		propertyHelp = new String[numProperties];
		propertyIdxMap = new int[numProperties];
		revPropertyIdxMap = new int[numProperties];

		activeProperty = -1;  // initialize for addProperty

		/* initialize propertyIdxMap to take care of legacy items */
		for (int i = 0; i < numProperties; i++) propertyIdxMap[i] = i;
		for (int i = 0; i < numProperties; i++) revPropertyIdxMap[i] = i;
	}

	public void reallocateElementNameList() {
		/* Reallocate the device name list to improve the performance of searches */
		elementNameList = null;  // throw away the old one
		elementNameList = new HashListImpl(2 * elementList.size());  // make a new one

		// do this using the names of the elements rather than the old list because it might be
		// messed up if an element gets renamed

		for (int i = 0; i < this.elementList.size(); i++)
			elementNameList.add( elementList.get(i).getName() );
	}

	private void resynchElementNameList() {
		reallocateElementNameList();
		elementNamesOutOfSynch = false;
	}

	public int getElementCount() {
		return elementList.size();
	}

	public int getNumProperties() {
		return numProperties;
	}

	public void setNumProperties(int num) {
		numProperties = num;
	}

	public String[] getPropertyName() {
		return propertyName;
	}

	public void setPropertyName(String[] name) {
		propertyName = name;
	}

	public String[] getPropertyHelp() {
		return propertyHelp;
	}

	public void setPropertyHelp(String[] help) {
		propertyHelp = help;
	}

	public int[] getPropertyIdxMap() {
		return propertyIdxMap;
	}

	public void setPropertyIdxMap(int[] map) {
		propertyIdxMap = map;
	}

	public int[] getRevPropertyIdxMap() {
		return revPropertyIdxMap;
	}

	public void setRevPropertyIdxMap(int[] map) {
		revPropertyIdxMap = map;
	}

	public int getDSSClassType() {
		return classType;
	}

	public void setDSSClassType(int type) {
		classType = type;
	}

	public ObjectList<T> getElementList() {
		return elementList;
	}

	public void setElementList(ObjectList<T> list) {
		elementList = list;
	}

	public boolean isElementNamesOutOfSynch() {
		return elementNamesOutOfSynch;
	}

	public void setElementNamesOutOfSynch(boolean value) {
		elementNamesOutOfSynch = value;
	}

	public boolean isSaved() {
		return saved;
	}

	public void setSaved(boolean value) {
		saved = value;
	}

	public String getName() {
		return className;
	}

	public static void setDSSClasses(com.epri.dss.common.DSSClasses classes) {
		DSSClasses = classes;
	}

	public static com.epri.dss.common.DSSClasses getDSSClasses() {
		return DSSClasses;
	}

}

File: pydss/common/DSSClassImpl.py
from dss.shared.impl.PointerListImpl import (PointerListImpl,)
from dss.common.DSSClass import (DSSClass,)
from dss.shared.impl.HashListImpl import (HashListImpl,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.parser.impl.Parser import (Parser,)


class DSSClassImpl(object, DSSClass):
    """Base class for all DSS collection classes.
    Keeps track of objects of each class, dispatches edits, etc
    """
    DSSClasses = None
    className = None
    activeElement = None
    commandList = None
    activeProperty = None
    elementNameList = None
    numProperties = None
    propertyName = None
    propertyHelp = None
    propertyIdxMap = None
    revPropertyIdxMap = None
    classType = None
    elementList = None
    elementNamesOutOfSynch = None
    saved = None

    def __init__(self):
        super(DSSClassImpl, self)()
        # init size and increment
        self.elementList = PointerListImpl(20)
        self.propertyName = None
        self.propertyHelp = None
        self.propertyIdxMap = None
        self.revPropertyIdxMap = None
        self.activeElement = -1
        self.activeProperty = -1
        self.elementNameList = HashListImpl(100)
        self.elementNamesOutOfSynch = False

    def finalize(self):
        # get rid of space occupied by strings
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.numProperties):
                break
            self.propertyName[i] = ''
        _1 = True
        i = 0
        while True:
            if _1 is True:
                _1 = False
            else:
                i += 1
            if not (i < self.numProperties):
                break
            self.propertyHelp[i] = ''
        self.propertyName = [None] * 0
        self.propertyHelp = [None] * 0
        self.propertyIdxMap = [None] * 0
        self.revPropertyIdxMap = [None] * 0
        self.elementList = None
        self.elementNameList = None
        self.commandList = None
        super(DSSClassImpl, self).finalize()

    def newObject(self, ObjName):
        DSSGlobals.doErrorMsg('Reached base class of DSSClass for device \"' + ObjName + '\"', 'N/A', 'Should be overridden.', 780)
        return 0

    def setActiveElement(self, Value):
        if Value > 0 and Value <= len(self.elementList):
            self.activeElement = Value
            DSSGlobals.activeDSSObject = self.elementList.get(self.activeElement)
            if isinstance(DSSGlobals.activeDSSObject, DSSCktElement):
                DSSGlobals.activeCircuit.setActiveCktElement(DSSGlobals.activeDSSObject)

    def getActiveElement(self):
        return self.activeElement

    def edit(self):
        """Uses global parser."""
        DSSGlobals.doSimpleMsg('DSSClass.edit() called. Should be overriden.', 781)
        return 0

    def init(self, Handle):
        DSSGlobals.doSimpleMsg('DSSClass.init() called. Should be overriden.', 782)
        return 0

    def addObjectToList(self, Obj):
        """Used by newObject()."""
        # put it in this collection's element list.
        self.elementList.add(Obj)
        self.elementNameList.add(Obj.getName())
        if len(self.elementList) > 2 * self.elementNameList.getInitialAllocation():
            self.reallocateElementNameList()
        self.setActiveElement(len(self.elementList) - 1)
        return self.activeElement
        # return index of object in list

    def setActive(self, ObjName):
        result = False
        if self.elementNamesOutOfSynch:
            self.resynchElementNameList()
        idx = self.elementNameList.find(ObjName)
        if idx > 0:
            self.setActiveElement(idx)
            DSSGlobals.activeDSSObject = self.elementList.get(idx)
            result = True
        return result

    def find(self, ObjName):
        """Find an obj of this class by name."""
        result = None
        if self.elementNamesOutOfSynch:
            self.resynchElementNameList()
        idx = self.elementNameList.find(ObjName)
        if idx > 0:
            self.setActiveElement(idx)
            result = self.elementList.get(idx)
        return result

    def getActiveObj(self):
        """Get address of active obj of this class."""
        if self.activeElement > 0:
            return self.elementList.get(self.activeElement)
        else:
            return None

    def getFirstPropertyName(self):
        self.activeProperty = 0
        return self.getNextPropertyName()

    def getNextPropertyName(self):
        if self.activeProperty <= self.numProperties:
            result = self.propertyName[self.activeProperty]
        else:
            result = ''
        self.activeProperty += 1
        return result

    def propertyIndex(self, prop):
        """Find property value by string."""
        result = 0
        # default result if not found
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.numProperties):
                break
            if prop.equalsIgnoreCase(self.propertyName[i]):
                result = self.propertyIdxMap[i]
                break
        return result

    def countProperties(self):
        """Add no. of intrinsic properties."""
        self.numProperties = self.numProperties + 1

    def defineProperties(self):
        """Add properties of this class to propName."""
        self.activeProperty += 1
        self.propertyName[self.activeProperty] = 'like'
        self.propertyHelp[self.activeProperty] = 'Make like another object, e.g.:' + DSSGlobals.CRLF + DSSGlobals.CRLF + 'new capacitor.C2 like=c1  ...'

    def classEdit(self, activeObj, paramPointer):
        # continue parsing with contents of parser
        if paramPointer > 0:
            # DSSObject obj = (DSSObject) ActiveObj;
            _0 = paramPointer
            _1 = False
            while True:
                if _0 == 1:
                    _1 = True
                    self.makeLike(Parser.getInstance().makeString())
                    # like command
                    break
                break
        return 0

    def makeLike(self, objName):
        DSSGlobals.doSimpleMsg('DSSClass.makeLike() called. Should be overriden.', 784)
        return 0

    def getFirst(self):
        result = 0
        if len(self.elementList) == 0:
            result = 0
        else:
            self.setActiveElement(1)
            DSSGlobals.activeDSSObject = self.elementList.get(0)
            if isinstance(DSSGlobals.activeDSSObject, DSSCktElement):
                DSSGlobals.activeCircuit.setActiveCktElement(DSSGlobals.activeDSSObject)
                result = self.activeElement
        return result

    def getNext(self):
        result = -1
        if self.activeElement > len(self.elementList):
            result = 0
        else:
            DSSGlobals.activeDSSObject = self.elementList.getNext()
            if isinstance(DSSGlobals.activeDSSObject, DSSCktElement):
                DSSGlobals.activeCircuit.setActiveCktElement(DSSGlobals.activeDSSObject)
                result = self.activeElement
        self.activeElement += 1
        return result

    def addProperty(self, propName, cmdMapIndex, helpString):
        """Helper routine for building property strings.

        Using the addProperty function, you can list the properties here in the order you want
        them to appear when properties are accessed sequentially without tags. Syntax:

        addProperty(<name of property>, <index in the edit case statement>, <help text>);
        """
        self.activeProperty += 1
        self.propertyName[self.activeProperty] = propName
        self.propertyHelp[self.activeProperty] = helpString
        # map to internal object property index
        self.propertyIdxMap[self.activeProperty] = cmdMapIndex
        self.revPropertyIdxMap[cmdMapIndex] = self.activeProperty

    def allocatePropertyArrays(self):
        self.propertyName = [None] * self.numProperties
        self.propertyHelp = [None] * self.numProperties
        self.propertyIdxMap = [None] * self.numProperties
        self.revPropertyIdxMap = [None] * self.numProperties
        self.activeProperty = -1
        # initialize for addProperty
        # initialize propertyIdxMap to take care of legacy items
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.numProperties):
                break
            self.propertyIdxMap[i] = i
        _1 = True
        i = 0
        while True:
            if _1 is True:
                _1 = False
            else:
                i += 1
            if not (i < self.numProperties):
                break
            self.revPropertyIdxMap[i] = i

    def reallocateElementNameList(self):
        # Reallocate the device name list to improve the performance of searches
        self.elementNameList = None
        # throw away the old one
        self.elementNameList = HashListImpl(2 * len(self.elementList))
        # make a new one
        # do this using the names of the elements rather than the old list because it might be
        # messed up if an element gets renamed
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < len(self.elementList)):
                break
            self.elementNameList.add(self.elementList.get(i).getName())

    def resynchElementNameList(self):
        self.reallocateElementNameList()
        self.elementNamesOutOfSynch = False

    def getElementCount(self):
        return len(self.elementList)

    def getNumProperties(self):
        return self.numProperties

    def setNumProperties(self, num):
        self.numProperties = num

    def getPropertyName(self):
        return self.propertyName

    def setPropertyName(self, name):
        self.propertyName = name

    def getPropertyHelp(self):
        return self.propertyHelp

    def setPropertyHelp(self, help):
        self.propertyHelp = help

    def getPropertyIdxMap(self):
        return self.propertyIdxMap

    def setPropertyIdxMap(self, map):
        self.propertyIdxMap = map

    def getRevPropertyIdxMap(self):
        return self.revPropertyIdxMap

    def setRevPropertyIdxMap(self, map):
        self.revPropertyIdxMap = map

    def getDSSClassType(self):
        return self.classType

    def setDSSClassType(self, type):
        self.classType = type

    def getElementList(self):
        return self.elementList

    def setElementList(self, list):
        self.elementList = list

    def isElementNamesOutOfSynch(self):
        return self.elementNamesOutOfSynch

    def setElementNamesOutOfSynch(self, value):
        self.elementNamesOutOfSynch = value

    def isSaved(self):
        return self.saved

    def setSaved(self, value):
        self.saved = value

    def getName(self):
        return self.className

    @classmethod
    def setDSSClasses(cls, classes):
        cls.DSSClasses = classes

    @classmethod
    def getDSSClasses(cls):
        return cls.DSSClasses
